### PR TITLE
Link preview generation

### LINKED ISSUE #11158: Add unfurling of URLs in chats
https://github.com/status-im/status/issues/2

### User Story
As a user, I want the app to unfurl URLs so that I can easily see what they refer to without needing to open the URL.

### Description
[comment]: # (Feature or Bug? i.e Type: Bug)
_Type_: Feature

[comment]: # (Describe the feature you would like, or briefly summarise the bug and what you did, what you expected to happen, and what actually happens. Sections below)
_Summary_: If another user shares a URL with me (maybe from a 3rd party app #4605), it would be great if I could see the summary of the content of the URL so I have a quick idea of what it's about.

#### Expected behavior
[comment]: # (Describe what you expected to happen.)

1. User receives URL
1. App recognizes URL and uses an unfurlingeloper-blog/everything-you-ever-wanted-to-know-about-unfurling-but-were-afraid-to-ask-or-how-to-make-your-e64b4bb9254) library to generate an image (or HTML) content that represents the summary of the URL.

image

### Acceptance Criteria

* [ ] Links starting with `http://` or `https://` are unfurled on the receiving client side only for the URLs in the whitelist if enabled
* [ ] Performance is not noticeably negatively impacted by the new functionality
* [ ] The whitelist includes https://github.com

### Security implications
We have to think about the security implications if we do this on the client side (will it be acceptable to access a random URL which might leak metadata about us?)

### PR DESCRIPTION
**Pull request by vkjr**
fixes #11158

### Summary

Link preview generation implemented. Currently works for `youtube`.

Screenshot 2020-10-20 at 16 13 18
Screenshot 2020-10-20 at 16 13 14
Screenshot 2020-10-20 at 16 13 04

#### Platforms

- Android
- iOS

##### Functional

- 1-1 chats
- public chats
- group chats

### Steps to test
- Open Status
- Navigate to chat with links

status: ready

### CODE REVIEW

**Comment by qoqobolo** [comment]
@vkjr there are conflicts in the branch. Could you take a look, please?

File: modules/react-native-status/android/src/main/java/im/status/ethereum/module/StatusModule.java
Comment (flexsurfer) [inline modules/react-native-status/android/src/main/java/im/status/ethereum/module/StatusModule.java]: just wondering why not a service ?

File: src/status_im/chat/models/link_preview.cljs
Comment (flexsurfer) [inline src/status_im/chat/models/link_preview.cljs]: `do` - leftover ?

File: src/status_im/chat/models/link_preview.cljs
Comment (flexsurfer) [inline src/status_im/chat/models/link_preview.cljs:28]: could you pls elaborate on this ?

File: src/status_im/constants.cljs
Comment (flexsurfer) [inline src/status_im/constants.cljs]: why is this option needed?

File: src/status_im/chat/models/link_preview.cljs
Comment (flexsurfer) [inline src/status_im/chat/models/link_preview.cljs:10]: `{:events [:link-preview/enable]}`

File: src/status_im/events.cljs
Comment (flexsurfer) [inline src/status_im/events.cljs]: we don't use this ns anymore, just use `{:events []}` in the `link-preview` ns

File: src/status_im/ui/screens/chat/message/message.cljs
Comment (flexsurfer) [inline src/status_im/ui/screens/chat/message/message.cljs]: can we move this to status-go, so we'll have links from the whitelist in `(:valid-links (:content message))`

**Review by flexsurfer — COMMENTED** [review]
overall looks great

File: src/status_im/ui/screens/chat/message/message.cljs
Comment (cammellos) [reply src/status_im/ui/screens/chat/message/message.cljs]: that's a good idea, maybe is worth to have only: `links` (say something like `{:domain ... :url ...}` rather than `valid-links`, and in status-react we match against a domain whitelist.
The rationale is the whitelist might be changing because the user enables/disables and we won't be reloading messages probably, what do you think?

Comment (flexsurfer) [reply src/status_im/ui/screens/chat/message/message.cljs]: yeah all links is better, i wanted to suggest it first, but for some reason changed my mind in the last moment :)

File: src/status_im/chat/models/link_preview.cljs
Comment (vkjr) [reply src/status_im/chat/models/link_preview.cljs]: yep

File: src/status_im/chat/models/link_preview.cljs
Comment (vkjr) [reply src/status_im/chat/models/link_preview.cljs:28]: During session we save locally data about link preview to not fetch it every time when chat re-creates

File: src/status_im/constants.cljs
Comment (vkjr) [reply src/status_im/constants.cljs]: will be renamed but it is set to `false` when user says "Don't ask me again about link preview"

File: src/status_im/ui/screens/chat/message/message.cljs
Comment (vkjr) [reply src/status_im/ui/screens/chat/message/message.cljs]: Thanks, will dig it on status-go side

Comment (cammellos) [reply src/status_im/ui/screens/chat/message/message.cljs]: https://github.com/status-im/status-go/blob/7467ca7b103f685dd64d73485555f90c181a4484/protocol/common/message.go#L306 is a class you can modify to fetch links and store them in the database, similarly to how we do with mentions

File: src/status_im/ethereum/json_rpc.cljs
Comment (flexsurfer) [inline src/status_im/ethereum/json_rpc.cljs:85]: probably better to move this to accounts or settings service

**Comment by vkjr** [comment]
@cammellos, @flexsurfer, review notes addressed, could you please check again?

@qoqobolo, rebased)

**Comment by qoqobolo** [comment]
@vkjr could you rebase it again please? :) The last rebase was done before  #11260 was merged, and this caused the tests to fail.

**Comment by vkjr** [comment]
@qoqobolo - done :)

**Comment by churik** [comment]
All failed e2e are not related to PR (ropsten network issues)

**Comment by churik** [comment]
@vkjr 
looks awesome! question - is it expected that long-press on preview is not working? 
 So to see the menu for reactions I need to long-tap on the main message with link.

**Comment by vkjr** [comment]
@churik, hmm, it was never discussed if reaction should work on the link-preview, so I can say it is expected since it wasn't implemented :))
@errorists, wdyt on this?

Anyway I'd be happy to merge this and add improvements as a separate prs :)

**Comment by churik** [comment]
also not sure about transparency of preview, it can cause such behavior:
IMAGE 2020-10-30 16:11:40

cc @errorists is it OK to merge like this from your POV?

**Comment by errorists** [comment]
whoa, I didn't know we had a PR for this already. cool 😅 the fix for transparency looks simple enough, just place a white background for the container. I think yes, long press and reactions should be possible for the unfurled message alone with actions: Open and Copy, without replies.

oh and yes it's ok to merge if we follow up with the long press menu fixes

**Comment by vkjr** [comment]
Would be happy implement long press menu fixes immediately after this one)

**Comment by churik** [comment]
@vkjr dark mode is affected
IMAGE 2020-10-30 17:01:33

**Comment by vkjr** [comment]
@churik, fixed :)

**Comment by churik** [comment]
Thanks @vkjr !
All good now.

**Comment by vkjr** [comment]
@cammellos, @flexsurfer, needs your approval :)

File: src/status_im/native_module/core.cljs
Comment (flexsurfer) [inline src/status_im/native_module/core.cljs]: leftover?

**Review by flexsurfer — APPROVED** [review]

File: src/status_im/native_module/core.cljs
Comment (vkjr) [reply src/status_im/native_module/core.cljs]: yes, thanks for catching!

**Pull request merged**
